File: source/MRMesh/MRId.cpp
#include "MRId.h"
#include "MRGTest.h"

namespace MR
{

TEST(MRMesh, Id) 
{
    EdgeId e(1);
    FaceId f(2);
    VertId v(3);

    int ei = e;
    int fi = f;
    int vi = v;

    EXPECT_EQ( e, ei );
    EXPECT_EQ( f, fi );
    EXPECT_EQ( v, vi );

    EXPECT_EQ( EdgeId().valid(), false );
    EXPECT_EQ( e.valid(), true );
}

} //namespace MR

File: source/MRDotNet/MRId.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;

namespace MR
{
    public partial class DotNet
    {
        [StructLayout(LayoutKind.Sequential)]
        public struct VertId
        {
            public int Id;
            public VertId(int id = -1) { Id = id; }
            public bool Valid() => Id >= 0;
        }

        [StructLayout(LayoutKind.Sequential)]
        public struct FaceId
        {
            public int Id;
            public FaceId(int id = -1) { Id = id; }
            public bool Valid() => Id >= 0;
        }

        [StructLayout(LayoutKind.Sequential)]
        public struct EdgeId
        {
            public int Id;
            public EdgeId(int id = -1) { Id = id; }
            public bool Valid() => Id >= 0;
        }


        [StructLayout(LayoutKind.Sequential)]
        public struct UndirectedEdgeId
        {
            public int Id;
            public UndirectedEdgeId(int id = -1) { Id = id; }
            public bool Valid() => Id >= 0;
        }

        [StructLayout(LayoutKind.Sequential)]
        public struct RegionId
        {
            public int Id;
            public RegionId(int id = -1) { Id = id; }
            public bool Valid() => Id >= 0;
        }
    }
}
